spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -2238,11 +2240,14 @@ protected void doInTransactionWithoutResult(TransactionStatus status) {
 
 		private List<ConsumerRecord<K, V>> createRecordList(final ConsumerRecords<K, V> records) {
 			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
-			List<ConsumerRecord<K, V>> list = new LinkedList<>();
+			@SuppressWarnings("unchecked") ConsumerRecord<K, V>[] recordsArray =
+					(ConsumerRecord<K, V>[]) Array.newInstance(ConsumerRecord.class, records.count());
+			int index = 0;
 			while (iterator.hasNext()) {
-				list.add(iterator.next());
+				recordsArray[index] = iterator.next();
+				index += 1;
 			}
-			return list;
+			return Arrays.asList(recordsArray);
```

Comment:
Why not `new ArrayList()`?

Reply:
`Arrays.asList` is

```java
  public static <T> List<T> asList(T... a) {
        return new ArrayList<>(a);
    }
```

unfortunately `recordsArray` is just an array here, and `ArrayList` constructor requires a `Collection`..

I'm not exactly sure how it works, but it seems that java does some magic to transform `T... a` into `Collection<T>`

Reply:
Right. Let's populate that `new ArrayList()` manually then!
Should not be technically too much difference with this `asList()` in performance.
I worry more about breaking change you made for `RecordFilterStrategy` and potential breaking changes in the target applications due to non-consistent result object behavior.